File: 1038-binary-search-tree-to-greater-sum-tree/1038-binary-search-tree-to-greater-sum-tree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
   
    private int sum = 0;
    
    public TreeNode BstToGst(TreeNode root) {
        Traverse(root);
        return root;
    }
    
    private void Traverse(TreeNode node) {
        if (node == null) return;
        
        // Traverse right subtree
        Traverse(node.right);
        
        // Update the current node's value
        sum += node.val;
        node.val = sum;
        
        // Traverse left subtree
        Traverse(node.left);
    }
}

File: 1038-binary-search-tree-to-greater-sum-tree/1038-binary-search-tree-to-greater-sum-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    int sum = 0;
    
    void Traverse(TreeNode* node) {
        if (node == nullptr) return;
        
        // Traverse right subtree
        Traverse(node->right);
        
        // Update the current node's value
        sum += node->val;
        node->val = sum;
        
        // Traverse left subtree
        Traverse(node->left);
    }
public:
    TreeNode* bstToGst(TreeNode* root) {
        Traverse(root);
        return root;
    }
};